[MLIR][Intrinsics] Add new MLIR API to automatically resolve overload types

rajatbajpai opened the pull request:
Add createIntrinsicCall overload that accepts return type and arguments, automatically resolve overload types rather than requiring manual computation. Simplifies NVVM_PrefetchOp by removing conditional overload logic.

durga4github commented on mlir/include/mlir/Dialect/LLVMIR/NVVMOps.td:
```diff
@@ -3176,12 +3176,7 @@ def NVVM_PrefetchOp : NVVM_Op<"prefetch",
-      createIntrinsicCall(builder, id, args, {args[0]->getType()});
-    else
-      createIntrinsicCall(builder, id, args);
+    createIntrinsicCall(builder, id, builder.getVoidTy(), args);
```
A general +1 from me for this enhancement. This can simplify (i.e. remove) the if-checks on a number of NVVM Ops.

jurahul commented on llvm/include/llvm/IR/Intrinsics.h:
```diff
@@ -104,6 +104,21 @@ namespace Intrinsic {
+  /// \param ArgTys - The argument types of the intrinsic.
+  LLVM_ABI Function *getOrInsertDeclaration(Module *M, ID IID, Type *RetTy,
+                                            ArrayRef<Type *> ArgTys);
+
```
Can you add a unit test for this API to llvm/unittest/IR/IntrinsicsTest.cpp ?

rajatbajpai replied:
Done

jurahul commented on llvm/include/llvm/IR/Intrinsics.h:
```diff
@@ -104,6 +104,21 @@ namespace Intrinsic {
+  /// \param IID - The intrinsic ID.
+  /// \param RetTy - The return type of the intrinsic.
+  /// \param ArgTys - The argument types of the intrinsic.
+  LLVM_ABI Function *getOrInsertDeclaration(Module *M, ID IID, Type *RetTy,
```
Is it possible to adopt this API in some existing LLVM/clang code to demonstrate better ergonomics? I understand that you are adopting this in MLIR LLVM dialect -> LLVM IR translation, so just curious

grypp commented on llvm/unittests/IR/IntrinsicsTest.cpp:
```diff
@@ -46,6 +44,8 @@ class IntrinsicsTest : public ::testing::Test {
     EXPECT_NE(BB, nullptr);
   }
 
+  void TearDown() override { M.reset(); }
```
accidental move?

rajatbajpai committed:
[MLIR][Intrinsic] Add createIntrinsicCall with automatic overload resolution

Add createIntrinsicCall overload that accepts return type and arguments,
automatically resolve overload types rather than requiring manual computation.
Simplifies NVVM_PrefetchOp by removing conditional overload logic.

rajatbajpai commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -898,6 +898,21 @@ llvm::CallInst *mlir::LLVM::detail::createIntrinsicCall(
+    llvm::Type *retTy, ArrayRef<llvm::Value *> args) {
+  llvm::Module *module = builder.GetInsertBlock()->getModule();
+
+  SmallVector<llvm::Type *> argTys;
```
Created a PR https://github.com/llvm/llvm-project/pull/169007 for LLVM side of things.

rajatbajpai commented:
@grypp @joker-eph Could you please review this PR? It has been separated from the LLVM changes and is now an independent change.

Dinistro reviewed (approved):
The other patch is now not strictly required, right?
If so, LGTM!

rajatbajpai commented:
Recent builds are passing.
